File: py/packages/genkit-ai/src/genkit/blocks/document.py
# Copyright 2025 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# SPDX-License-Identifier: Apache-2.0

"""Document model for the Genkit framework.

This module provides types for creating and managing multi-modal documents in
Genkit.
"""

from __future__ import annotations

from copy import deepcopy
from typing import Any

from genkit.core.typing import (
    DocumentData,
    DocumentPart,
    Embedding,
    Media,
)

TEXT_DATA_TYPE: str = 'text'


class Document(DocumentData):
    """Represents document content along with its metadata.

    This object can be embedded, indexed or retrieved. Each document can contain
    multiple parts (for example text and an image).
    """

    def __init__(
        self,
        content: list[DocumentPart],
        metadata: dict[str, Any] | None = None,
    ) -> None:
        """Initializes a Document object.

        Performs a deep copy of the provided content and metadata to prevent
        unintended modifications to the original objects.

        Args:
            content: A list of DocumentPart objects representing the document's content.
            metadata: An optional dictionary containing metadata about the document.
        """
        doc_content = deepcopy(content)
        doc_metadata = deepcopy(metadata)
        super().__init__(content=doc_content, metadata=doc_metadata)

    @staticmethod
    def from_document_data(document_data: DocumentData) -> Document:
        """Constructs a Document instance from a DocumentData object.

        Args:
            document_data: The DocumentData object containing content and metadata.

        Returns:
            A new Document instance initialized with the provided data.
        """
        return Document(content=document_data.content, metadata=document_data.metadata)

    @staticmethod
    def from_text(text: str, metadata: dict[str, Any] | None = None) -> Document:
        """Constructs a Document instance from a single text string.

        Args:
            text: The text content for the document.
            metadata: Optional metadata for the document.

        Returns:
            A new Document instance containing a single text part.
        """
        return Document(content=[DocumentPart(text=text)], metadata=metadata)

    @staticmethod
    def from_media(
        url: str,
        content_type: str | None = None,
        metadata: dict[str, Any] | None = None,
    ) -> Document:
        """Constructs a Document instance from a single media URL.

        Args:
            url: The URL of the media content.
            content_type: Optional MIME type of the media.
            metadata: Optional metadata for the document.

        Returns:
            A new Document instance containing a single media part.
        """
        return Document(
            content=[DocumentPart(media=Media(url=url, content_type=content_type))],
            metadata=metadata,
        )

    @staticmethod
    def from_data(
        data: str,
        data_type: str | None = None,
        metadata: dict[str, Any] | None = None,
    ) -> Document:
        """Constructs a Document instance from a single data string, determining type.

        If `data_type` is 'text', creates a text document. Otherwise, assumes
        `data` is a URL and creates a media document.

        Args:
            data: The data string (either text content or a media URL).
            data_type: The type of the data ('text' or a media content type).
            metadata: Optional metadata for the document.

        Returns:
            A new Document instance.
        """
        if data_type == TEXT_DATA_TYPE:
            return Document.from_text(data, metadata)
        return Document.from_media(data, data_type, metadata)

    def text(self) -> str:
        """Concatenates all text parts of the document's content.

        Returns:
            A single string containing the text from all text parts, joined
            without delimiters.
        """
        return ''.join(p.root.text if p.root.text is not None else '' for p in self.content)

    def media(self) -> list[Media]:
        """Retrieves all media parts from the document's content.

        Returns:
            A list of Media objects contained within the document.
        """
        media_parts = [part.root.media for part in self.content]
        return list(filter(lambda m: m is not None, media_parts))

    def data(self) -> str:
        """Gets the primary data content of the document.

        Returns the concatenated text if available, otherwise the URL of the
        first media part. Returns an empty string if the document has neither
        text nor media.

        Returns:
            The primary data string (text or media URL).
        """
        if self.text():
            return self.text()

        if self.media():
            return self.media()[0].url

        return ''

    def data_type(self) -> str | None:
        """Gets the data type corresponding to the primary data content.

        Returns 'text' if the primary data is text. If the primary data is media,
        returns the content type of the first media part. Returns None if the
        document has no primary data or the media content type is not set.

        Returns:
            The data type string ('text' or a MIME type) or None.
        """
        if self.text():
            return TEXT_DATA_TYPE

        if self.media() and self.media()[0].content_type:
            return self.media()[0].content_type

        return None

    def get_embedding_documents(self, embeddings: list[Embedding]) -> list[Document]:
        """Creates multiple Document instances from a single document and its embeddings.

        Since embedders can return multiple embeddings for one input document,
        but storage often requires a 1:1 document-to-embedding relationship,
        this method duplicates the original document for each embedding.
        Embedding metadata is added to the corresponding document's metadata
        under the 'embedMetadata' key.

        Args:
            embeddings: A list of Embedding objects generated for this document.

        Returns:
            A list of Document objects, one for each provided embedding.
        """
        documents = []
        for embedding in embeddings:
            content = deepcopy(self.content)
            metadata = deepcopy(self.metadata)
            if embedding.metadata:
                if not metadata:
                    metadata = {}
                metadata['embedMetadata'] = embedding.metadata
            documents.append(Document(content=content, metadata=metadata))
        check_unique_documents(documents)
        return documents


def check_unique_documents(documents: list[Document]) -> bool:
    """Checks if a list of documents contains duplicates based on their JSON representation.

    Prints a warning if duplicates are found, as this can cause issues with
    vector storage systems that key documents by their content hash.
    Duplicate documents might arise if embeddings are generated without unique
    metadata for different parts or aspects of the same original document.

    Args:
        documents: A list of Document objects to check.

    Returns:
        True if all documents are unique, False otherwise (primarily for testing).
    """
    seen = set()
    for doc in documents:
        if doc.model_dump_json() in seen:
            print(
                """
                Warning: embedding documents are not unique.
                Are you missing embed metadata?
                """
            )
            return False
        seen.add(doc.model_dump_json())
    return True

File: java/ai/src/main/java/com/google/genkit/ai/Document.java
/*
 * Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package com.google.genkit.ai;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;

/**
 * Document represents a document for use with embedders and retrievers.
 */
@JsonInclude(JsonInclude.Include.NON_NULL)
public class Document {

  @JsonProperty("content")
  private List<Part> content;

  @JsonProperty("metadata")
  private Map<String, Object> metadata;

  /**
   * Default constructor.
   */
  public Document() {
    this.content = new ArrayList<>();
    this.metadata = new HashMap<>();
  }

  /**
   * Creates a Document with text content.
   *
   * @param text
   *            the text content
   */
  public Document(String text) {
    this();
    this.content.add(Part.text(text));
  }

  /**
   * Creates a Document with parts.
   *
   * @param content
   *            the content parts
   */
  public Document(List<Part> content) {
    this.content = content != null ? content : new ArrayList<>();
    this.metadata = new HashMap<>();
  }

  /**
   * Creates a text Document.
   *
   * @param text
   *            the text content
   * @return a Document with text content
   */
  public static Document fromText(String text) {
    return new Document(text);
  }

  /**
   * Creates a Document with text and metadata.
   *
   * @param text
   *            the text content
   * @param metadata
   *            the metadata
   * @return a Document with text content and metadata
   */
  public static Document fromText(String text, Map<String, Object> metadata) {
    Document doc = new Document(text);
    doc.metadata = metadata != null ? metadata : new HashMap<>();
    return doc;
  }

  /**
   * Gets the text content of this Document.
   *
   * @return the concatenated text content
   */
  public String text() {
    if (content == null) {
      return "";
    }
    StringBuilder sb = new StringBuilder();
    for (Part part : content) {
      if (part.getText() != null) {
        sb.append(part.getText());
      }
    }
    return sb.toString();
  }

  // Getters and setters

  public List<Part> getContent() {
    return content;
  }

  public void setContent(List<Part> content) {
    this.content = content;
  }

  public Map<String, Object> getMetadata() {
    return metadata;
  }

  public void setMetadata(Map<String, Object> metadata) {
    this.metadata = metadata;
  }

  /**
   * Adds metadata to this Document.
   *
   * @param key
   *            the metadata key
   * @param value
   *            the metadata value
   * @return this Document for chaining
   */
  public Document withMetadata(String key, Object value) {
    if (this.metadata == null) {
      this.metadata = new HashMap<>();
    }
    this.metadata.put(key, value);
    return this;
  }
}
